flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -60,49 +71,41 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
 		while (client == null) {
 			entry = clients.get(connectionId);
```

Comment:
I found it to be the cause of deadlock:
1. entry is placed by A
1. entry is read by B
1. A gets the connection and replaces entry - but A already holds an old entry
1. B acquires lock and waits for channel indefinitely

Reply:
Yes, you are right. Usually the exception can be notified to the old connecting channel by bootstrap. The test case actually triggers the exception within the connect function, thus we need to notify the exception explicitly.